lib/elixir/lib/regex.ex

```diff
@@ -1018,6 +1018,7 @@ defmodule Regex do
           {:ok, exported} = :re.compile(regex.source, [:export] ++ regex.opts)
 
           quote do
+            __assert_assert_no_match_or_guard_scope__("an escaped Regex")
             :re.import(unquote(Macro.escape(exported)))
```

Comment:
Instead of generating additional code, we should call `Regex.__import__(unquote(...))`, where `__import__` itself is a macro which also checks the environment.

Reply:
This sounds much better indeed 👍

Reply:
Arg, this might actually not work, since unlike `Kernel` we need to `require Regex`.

```elixir
       quote do
            require Regex
            Regex.__import_pattern__(unquote(Macro.escape(exported)))
          end
```

Which then makes it impossible to compile elixir:

```elixir
==> elixir (compile)
error: you are trying to use/import/require the module Regex which is currently being defined.
```

(although it seems weird to warn here, since it's happening within `quote` 🤔)

Reply:
I believe the issue happens because we escape the struct defined by `defstruct` inside Regex. An easy fix is to add this clause to the top of the `cond` in `__escape__`:

```
        is_nil(regex.re_pattern) ->
          nil
```

And then it compiles fine!

Reply:
Arf, it doesn't work either `invalid pattern in match, require is not allowed in matches`.

Reply:
TIL about the `required: true` meta, I think it fixed it nicely [0e671f4](https://github.com/elixir-lang/elixir/pull/14780/commits/0e671f42247d3b1b0642238dcf1836004de8e4d7).